R/parse_variable_indices.R

```diff
@@ -0,0 +1,148 @@
+#' Parse the indices from a vector of variables (extracted e.g. from a 
+#' `draws_summary` object or a `draws` object)
+#' 
+#' @name variable_indices
+#' @param x a character vector of variables
+#'
+#' @return
+#' A list with index information for each unique variable name V in `x`. Top-level list names are 
+#' the variable names. Each element contains: 
+#'    $dimensionality  the number of dimensions of V. Returns 0 for scalars with no brackets
+#'    but 1 for `y[1]` even if `y` has no other entries in `x`.
+#'    
+#'    $dimensions  a vector of the actual dimensions of V, as determined by the number of unique
+#'    elements at each index position. Set to `NA` if the dimensionality is zero.
+#'    
+#'    $implied_dimensions  a vector of the implied dimensions of V, where any position in V that 
+#'    contains exclusively integers is filled in to include all integers from the lesser of one 
+#'    and its minimum up to its maximum. Set to `NA` if the dimensionality is zero.
+#'    
+#'    $index_names  a list of length corresponding to the dimensionality, where each element is the
+#'    unique levels of the corresponding index if the index is parsed as factor, and NULL otherwise.
+#'    Set to `NULL` if the dimensionality is zero.
+#'    
+#'    $indices  if dimensionality is zero, returns 1.
+#'    if dimensionality is 1 or greater, returns a dataframe of every implied combination of indices
+#'    
+#'    $position  the position of each combination of indices from $indices in the the argument `x`
+#' 
+#' @details
+#' Assumes that variable indexing uses square brackets in the variable names
+#' 
+NULL
+
+parse_variable_indices <- function(x){
+  vars_indices <- strsplit(x, "(\\[|\\])")
+  vars <- sapply(vars_indices, `[[`, 1)
+  var_names <- unique(vars)
+  # Check that no variables contain unpaired or non-terminal square brackets
+  if_indexed <- (sapply(vars_indices, length) > 1)
+  if_indexed2 <- grepl("\\[.*\\]$", x)
+  extra_bracket <- grepl("\\].", x) | grepl("\\[.*\\[", x)
+  if (any(extra_bracket) | (!identical(if_indexed, if_indexed2))) {
+    stop_no_call(paste("Some variable names contain unpaired square brackets",
+                  "or are multi-indexed."))
+  }
+  missing_index <- grepl("\\[,|,\\]|,,", x) # check for `[,` `,,` or `,]`
```

Comment:
I would roll this into the regex above instead of making a specific check for it. Given that in most use cases these variable names are auto-generated by model code, I doubt that missing indices will be common enough to warrant time spent on an extra check rather than just having the index fail to parse. (Others might disagree though)

Reply:
done